lib/elastic_apm/central_config.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/central_config/cache_control'
+
+module ElasticAPM
+  # @api private
+  class CentralConfig
+    include Logging
+
+    # @api private
+    class ResponseError < InternalError
+      def initialize(response)
+        @response = response
+      end
+
+      attr_reader :response
+    end
+    class ClientError < ResponseError; end
+    class ServerError < ResponseError; end
+
+    DEFAULT_MAX_AGE = 300
+
+    def initialize(config)
+      @config = config
+      @modified_options = {}
+      @service_info = {
+        'service.name': config.service_name,
+        'service.environment': config.environment
+      }.to_json
+    end
+
+    attr_reader :config, :task
+
+    def start
+      return unless config.central_config?
+
+      fetch_and_apply_config
+    end
+
+    def fetch_and_apply_config
+      @task&.cancel
+      @task =
+        Concurrent::Promise
+        .execute(&method(:fetch_config))
+        .on_success(&method(:handle_success))
+        .rescue(&method(:handle_error))
+    end
+
+    def stop
+      @task&.cancel
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def fetch_config
+      resp = perform_request
+
+      case resp.status
+      when 200..299
+        resp
+      when 300..399
+        resp
+      when 400..499
+        raise ClientError, resp
+      when 500..599
+        raise ServerError, resp
+      end
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    def assign(update)
+      update.each_key do |key|
```

Comment:
I think this could do with some comments. Let me see if I'm getting it right:

- when a config attribute wasn't previously defined remotely, but now is, we'll record the current local value in `@modified_options`
- when a config attribute was previously defined remotely, but no longer is, we'll remove the value from `@modified_options` and revert to it

Is it possible for the local config to be updated after the agent is started?

Reply:
That's correct. It is technically possible to update some if not most of them on the go but I don't suspect anyone does it nor is it described nor encouraged anywhere.

Ruby being Ruby; it is possible.